File: aoc-2015/src/test/java/net/bbelovic/adventofcode/day2/Day2PuzzleMainTest.java
package net.bbelovic.adventofcode.day2;

import net.bbelovic.adventofcode.InputReader;
import net.bbelovic.adventofcode.Puzzle;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;

class Day2PuzzleMainTest {
    private static final InputReader INPUT_READER = new InputReader();
    private static final int DAY2_PUZZLE_PART1_RESULT = 1598415;
    private static final int DAY2_PUZZLE_PART2_RESULT = 3812909;

    @ParameterizedTest
    @MethodSource("testData")
    void should_solve_puzzle(Puzzle<String, Long> puzzle, long expectedResult) {
        long actualResult = INPUT_READER.readAllLines("input2.txt")
                .stream()
                .mapToLong(puzzle::solve)
                .sum();
        assertEquals(expectedResult, actualResult);
    }

    private static Stream<Arguments> testData() {
        return Stream.of(
                Arguments.of(new Day2PuzzlePart1(), DAY2_PUZZLE_PART1_RESULT),
                Arguments.of(new Day2PuzzlePart2(), DAY2_PUZZLE_PART2_RESULT)
        );
    }
}

File: aoc-2018/src/test/kotlin/net/bbelovic/adventofcode/year2018/day2/Day2PuzzleMainTest.kt
package net.bbelovic.adventofcode.year2018.day2

import net.bbelovic.adventofcode.InputReader
import net.bbelovic.adventofcode.Puzzle
import net.bbelovic.adventofcode.utils.dsl.testDataDsl
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource

const val EXPECTED_RESULT_PART_1 = 8715L
const val EXPECTED_RESULT_PART_2 = "fvstwblgqkhpuixdrnevmaycd"

class Day2PuzzleMainTest {
    @ParameterizedTest
    @MethodSource("testData")
    fun `should solve puzzle`(puzzle: Puzzle<List<String>, *>, expectedResult: Any) {
        val inputLines = InputReader().readAllLines("input2.txt")
        val actualChecksum = puzzle.solve(inputLines)
        assertEquals(expectedResult, actualChecksum)
    }

    companion object {
        @JvmStatic
        fun testData() = testDataDsl {
            testArguments (Day2PuzzlePart1(), EXPECTED_RESULT_PART_1)
            testArguments (Day2PuzzlePart2(), EXPECTED_RESULT_PART_2)
        }
    }
}